### PR TITLE
Improve rendering of surfaces by sorting surfaces by age and level an…

### PR DESCRIPTION
**Pull request by HavardNJ**
…d better base matching, improve label rendering on areas which are behind other by using topmost surface not rendered as base for label and resorting to iterating surfaces to find position/direction if the closest surface doesn't have values in the area where the label will be placed.

### CODE REVIEW

**Review by Brynjulf — APPROVED** [review]
LGTM

File: src/datautils/surfacedata.ts [hunk 1]
```diff
@@ -118,6 +64,99 @@ function getSurfaceLines(mappedSurfaces: any, trajectory: number[][]): SurfaceLi
...
+      }
+      return isValid;
+    })
+    .map((g: any, i: number) => {
+      const surface = surfaceAreas[g.name];
+      const top = surface[0];
+      return {
+        id: g.name,
+        color: unassignedColorScale(i),
+        top: top.data.map((d: number[]) => d[1]),
```
Comment (nilsml) [inline src/datautils/surfacedata.ts:96]: To make this code a little bit easier to understand, creating lamda functions of the content of the filter and map functions could be a good idea. 
`someArray.sort(onAge).filter(onlyValidGroups).map(toType)` or similar is easy to read.

Also I can see the use of `any` quite a lot in the code. Surly we know the types of most of these variables?!

File: src/datautils/surfacedata.ts [hunk 2]
```diff
@@ -177,53 +218,44 @@ const unassignedColorScale = scaleOrdinal<number, string>()
...
+  return null;
+}
 
-  if (scores.length < 2) {
-    return nextIndex;
-  }
-  scores.filter((s: any) => s.scores > 0).sort((a: any, b: any) => b.score - a.score || a.ci - b.ci);
-
-  return scores[0].ci;
+function isAnchestor(descendant: any, candidate: any, stratColumn: any): boolean {
```
Comment (nilsml) [inline src/datautils/surfacedata.ts:246]: Lots of `any`'s. Don't we know the type?

File: src/layers/GeomodelLabelsLayer.ts [hunk 3]
```diff
@@ -46,9 +50,40 @@ export class GeomodelLabelsLayer extends CanvasLayer {
...
+        return acc;
+      }
+      let topSum = 0;
+      let count = 0;
+      a.data.forEach((d: any) => {
+        if (d[1] != null) {
+          topSum += d[1];
+          count++;
+        }
+      });
```
Comment (nilsml) [inline src/layers/GeomodelLabelsLayer.ts:71]: Perhaps use reduce instead?

File: src/layers/GeomodelLabelsLayer.ts [hunk 4]
```diff
@@ -260,15 +333,45 @@ export class GeomodelLabelsLayer extends CanvasLayer {
...
+              const altSurface = surfaces[si++];
+              altY = findSampleAtPos(
+                altSurface.data.map((d: any) => [d[0], d[1]]),
+                x,
+                topLimit,
+                bottomLimit,
+              );
+            }
+          }
+        }
```
Comment (nilsml) [inline src/layers/GeomodelLabelsLayer.ts:371]: Quite a lot of nesting going on here. How about splitting up into small functions and try to reduce levels of nesting?

**Review by nilsml — COMMENTED** [review]
Quite a lot going on in this PR. Small PR's are easier to review. It might make sense to do it all in a big chunk, but this is a general statement.

A few things I have noticed in the code that I believe can be improved:

- The use of `any` type. I might not be the biggest Typescript fan, but since we are actually using Typescript I think we should try to be specific about types when we can
- Some of this code is quite difficult to understand for a novice like me. By splitting up code into smaller functions with descriptive names i think reading the code would be easier.
- 3-4 levels of nesting is too much in my opinion. Inversion of conditions, early returns, functional "patterns" or similar could be used to avoid this, which also leads to code that is easier to read and understand.

File: src/datautils/surfacedata.ts (hunk 1 above)
Comment (HavardNJ) [reply src/datautils/surfacedata.ts:96]: I'm kind of torn on this. I know "Clean code"(as in the book) prefers functions and has a hard limit of max four lines per function to enforce this (you can configure this as a rule in ESLint f.ex.). I find it easier to read this as sequential code rather than alot of small functions which fragments the code, atleast if you are interested in the nitty gritty details rather than getting an overview of what it does. Sometimes I feel like splitting it up into functions are like reading a book where you have to read footnotes for every sentence. But of course there are limits to how much code you can/should have in a function (thought I think 4 lines is too strict) and there is the added benefit of function name documenting intent.

Comment (HavardNJ) [reply src/datautils/surfacedata.ts:96]: As for any, at one point we decided that we provide types for the users of the component, but would not be strict with types internally. In surface data we combine and transform data and have quite a few temporary structures. I'm not sure if defining these types will be worth the effort.

File: src/datautils/surfacedata.ts (hunk 2 above)
Comment (HavardNJ) [reply src/datautils/surfacedata.ts:246]: I'll look into it

File: src/layers/GeomodelLabelsLayer.ts (hunk 3 above)
Comment (HavardNJ) [reply src/layers/GeomodelLabelsLayer.ts:71]: doh

File: src/layers/GeomodelLabelsLayer.ts (hunk 4 above)
Comment (HavardNJ) [reply src/layers/GeomodelLabelsLayer.ts:371]: I'll have a look at this

Comment (HavardNJ) [reply src/layers/GeomodelLabelsLayer.ts]: Cleaned up

File: src/layers/GeomodelLabelsLayer.ts (hunk 3 above)
Comment (HavardNJ) [reply src/layers/GeomodelLabelsLayer.ts]: Changed

File: src/datautils/surfacedata.ts (hunk 2 above)
Comment (HavardNJ) [reply src/datautils/surfacedata.ts]: Replaced a number of anys where we actually have the type

File: src/datautils/surfacedata.ts (hunk 1 above)
Comment (HavardNJ) [reply src/datautils/surfacedata.ts]: Cleaned up use of any

Comment (nilsml) [reply src/datautils/surfacedata.ts]: Better.
I still think splitting up into smaller functions are better unless understanding every bit of detail is important.
Most of the time it is ok to understand things on a higher level, and then dive into details if needed.
I guess we are discussing one of the main differences between imperative and functional programming, where functional programming is more about what and imperative is more about how. Different ways of writing and reading code.

File: src/layers/GeomodelLabelsLayer.ts (hunk 3 above)
Comment (nilsml) [reply src/layers/GeomodelLabelsLayer.ts]: A comment regarding this function; the input parameter to the reduce function is mutated making it impure. It is good practice to make reducers pure. Here is a quick example in js of how to make it a little bit more functional and pure:
```
const maybeApply = x => f => o => x ? f(o, x) : o;
const initialObject = { topSum: 0, count: 0, avgTopDepth: null, };
  
const reducer = (a, d) => maybeApply(d[1])((o, x) => ({ 
  topSum: o.topSum + x,
  count: o.count + 1,
  avgTopDepth: (o.topSum + x) / (o.count + 1),
}))(a);

const { avgTopDepth } = area.data.reduce(reducer, initialObject);
```

Comment (HavardNJ) [reply src/layers/GeomodelLabelsLayer.ts]: I agree with you in principle, but in this case the data structure of the accumulator is extremely simple and only lives in the reducer except for the single numeric value returned. Also I'm not sure that const maybeApply = x => f => o => x ? f(o, x) : o; and the calling structure is more readable than if (d[1] != null) for whoever is going to maintain this some time in the future. I also believe the test should be x != null as x = 0 would not be counted into the average. 
If there was a strong goal to not mutate the accumulator a simple alternative would be
(a, d) => {
   if (d != null) {
      return{
          topSum: a.topSum + d,
          count: a.count + 1,
          avgTopDepth: (a.topSum + d) / (a.count + 1),
      };
   }
   return a;
};
However this comes with the penalty of creating a new object for every value we iterate over. I tested this in a fiddle and it's about 50% slower than the reducer which reuses the accumulator over a million values. I also tested you suggestion which in the fiddle was roughly twice as slow as the one which reuse the accumulator. I'll add that the timing vary alot but the pattern was clear.

Comment (nilsml) [reply src/layers/GeomodelLabelsLayer.ts]: Performance wise you are correct of course. Creating a new object for each iteration is slower, especially in a language like Javascript. So if we are dealing with a really large number of iterations, it is not the best solution. In my experience this is rarely the case in frontend apps, so I don´t usually worry about it. But this is not your everyday frontend app I suppose...

Regarding my function replacing the if. The idea is to create a collection of this type of functions and put them somewhere where they are globally available. Then, when you use them a lot, you know what they do and don´t have to look at the code. It will be part of your standard library. Then you can get rid of many if´s and switch-statements which I personally think pollutes the code.

Comment (HavardNJ) [reply src/layers/GeomodelLabelsLayer.ts]: In many cases I agree, we should not do premature optimization and in most(all?) cases it is far more important to not mutate input data. But it's also a matter of how much effort you put into optimization and keeping performance in the back of your head. In this case we know we are iterating over some thousands values for each surface, the optimization effort is insignificant and we're sure that what we mutate is not used anywhere else . Given that we have a goal of opening a wellbore or field in <1sec in rep and fetching data takes the better part of that second(if not more), 5ms here and 20ms there quickly add up. (I might add that I stupidly put this code inside the drawing loop at first where it certainly would have an impact.)
I still think the maybeApply  = a => b => c => ... and maybeApply(...)(...)(...) are harder to read and understand, especially for novice programmers which may not have seen such constructs before, even if they are part of an internal lib. An alternative might be something like
const checkValid = d => (d != null);   
    const calcAvg= (a, d) => {
      	a.topSum += d;
      	++a.count;
      	a.avgTopDepth = a.topSum / a.count;
        return a;
    };
const { avgTopDepth } = data.reduce((a, d) => (checkValid(d[1]) ? calcAvg(a, d[1]) : a), {...initialObject});
but even this comes with a performance penalty suggesting that function calls (unsurprisingly?) impact performance 
Another option might be 
const { avgTopDepth } = data.filter(checkValid).reduce(calcAvg, {...initialObject});
however this is three to four(!!!) times slower in my test. (Interestingly, the original forEach loop is slower than the current reducer which is kind of puzzling though access to function parameters might be faster than accessing other variables).
IMHO the four lines of code as is does the job, is satisfactorily easy to read and performant. It'll do. There are better candidates for cleanup and refactoring in the component.

**Comment by Brynjulf** [comment]
Merging this as I believe it satisfies the condition of improving code health
